pkg/resources/management.cattle.io/v3/roletemplate/validator.go

```diff
@@ -73,65 +81,125 @@ func (a *admitter) Admit(request *admission.Request) (*admissionv1.AdmissionResp
+		return admission.ResponseBadRequest(err.Error()), nil
 	}
 
-	allowed, err := auth.EscalationAuthorized(request, roleTemplateGVR, a.sar, "")
+	allowed, err := auth.EscalationAuthorized(request, gvr, a.sar, "")
 	if err != nil {
 		logrus.Warnf("Failed to check for the 'escalate' verb on RoleTemplates: %v", err)
+	} else if allowed {
+		return admission.ResponseAllowed(), nil
 	}
 
-	if allowed {
-		return &admissionv1.AdmissionResponse{Allowed: true}, nil
+	err = auth.ConfirmNoEscalation(request, rules, "", a.resolver)
+	if err != nil {
+		return admission.ResponseFailedEscalation(err.Error()), nil
 	}
-	response := &admissionv1.AdmissionResponse{}
-	auth.SetEscalationResponse(response, auth.ConfirmNoEscalation(request, rules, "", a.resolver))
-	return response, nil
+
+	return admission.ResponseAllowed(), nil
+}
+
+// validateUpdateFields checks if the fields being changed are valid update fields.
+func (a *admitter) validateUpdateFields(oldRole, newRole *v3.RoleTemplate, fldPath *field.Path) *field.Error {
+	if err := validateContextValue(newRole, fldPath); err != nil {
+		return err
+	}
+
+	// if this is not a built in role no further validation is needed
+	if !oldRole.Builtin {
+		return nil
+	}
+
+	// allow changes to meta data and defaults
+	oldRole.ClusterCreatorDefault = newRole.ClusterCreatorDefault
+	oldRole.ProjectCreatorDefault = newRole.ProjectCreatorDefault
+
+	// we do not want to block K8s controllers from adding metadata or typemeta to the object.
+	oldRole.ObjectMeta = newRole.ObjectMeta
+	oldRole.TypeMeta = newRole.TypeMeta
+
+	if !equality.Semantic.DeepEqual(oldRole, newRole) {
+		return field.Forbidden(fldPath, "updates to builtIn RoleTemplates for fields other than CreatorDefault are forbidden")
+	}
+	return nil
+}
+
+// validateCreateFields checks if all required fields are present and valid.
+func validateCreateFields(newRole *v3.RoleTemplate, fldPath *field.Path) *field.Error {
+	return validateContextValue(newRole, fldPath)
+}
+
+func validateContextValue(newRole *v3.RoleTemplate, fldPath *field.Path) *field.Error {
+	if newRole.Administrative && newRole.Context != clusterContext {
+		return field.Forbidden(fldPath.Child("administrative"), "only cluster roles can be administrative")
+	}
+	if newRole.Context != clusterContext && newRole.Context != projectContext && newRole.Context != emptyContext {
+		return field.NotSupported(fldPath.Child("context"), newRole.Context, []string{clusterContext, projectContext})
+	}
+	return nil
+}
+
+func (a *admitter) validateDelete(oldRT *v3.RoleTemplate) (*admissionv1.AdmissionResponse, error) {
+	allRTs, err := a.roleTemplateResolver.RoleTemplateCache().List(labels.Everything())
+	if err != nil {
+		return nil, fmt.Errorf("failed to list all RoleTemplates: %w", err)
+	}
+
+	// verify that the role is not currently inherited
+	for _, roleTemplate := range allRTs {
+		for _, templateName := range roleTemplate.RoleTemplateNames {
+			if oldRT.Name == templateName {
```

Comment:
I wonder if an index could be created for this, something like the role binding indexer here https://github.com/rancher/webhook/blob/9d919f43a3d2c80a94ffe1bb222c3eb22ebe94cf/pkg/resources/core/v1/secret/mutator.go#L46